modules/internal/String.chpl

```diff
@@ -736,30 +899,56 @@ module String {
       return codepointToString(this.codepoint(idx));
     }
 
+    /*
+      Index into a string
+
+      :returns: A string with the complete multibyte character starting at the
+                specified byte index from ``1..string.length``
+     */
+    inline proc this(i: int) : string {
+      return this[i: codepointIndex];
+    }
+
     // Checks to see if r is inside the bounds of this and returns a finite
     // range that can be used to iterate over a section of the string
     // TODO: move into the public interface in some form? better name if so?
     pragma "no doc"
-    proc _getView(r:range(?)) where r.idxType != codepointIndex {
+    proc _getView(r:range(?)) where r.idxType == byteIndex {
       if boundsChecking {
         if r.hasLowBound() && (!r.hasHighBound() || r.size > 0) {
-          if r.low <= 0 then
+          if r.low:int <= 0 then
             halt("range out of bounds of string");
         }
         if r.hasHighBound() && (!r.hasLowBound() || r.size > 0) {
-          if (r.high < 0) || (r.high:int > this.len) then
+          if (r.high:int < 0) || (r.high:int > this.len) then
             halt("range out of bounds of string");
         }
       }
-      const ret = r[1:r.idxType..#(this.len:r.idxType)];
-      return ret;
+      const r1 = r[1:r.idxType..this.len:r.idxType];
+      if r1.stridable {
+        const ret = r1.low:int..r1.high:int by r1.stride;
+        return ret;
+      } else {
+        const ret = r1.low:int..r1.high:int;
+        return ret;
+      }
     }
 
     // Checks to see if r is inside the bounds of this and returns a finite
     // range that can be used to iterate over a section of the string.
     // Converts from codepointIndex range to byte index range in the process.
+    //
+    // Slicing by stridable codepoint ranges is unsupported because it
+    // creates an irregular sequence of bytes.  We could add support in the
+    // future by refactoring the callers of _getView() to add a slow path,
+    // or by storing an array of indices marking the beginning of each
+    // codepoint alongside the string.
     pragma "no doc"
-    proc _getView(r:range(?)) where r.idxType == codepointIndex {
+    proc _getView(r:range(?)) where r.idxType != byteIndex {
+      if r.stridable {
+        HaltWrappers.unimplementedFeatureHalt("string slicing",
```

Comment:
This should probably be a compile-time error (`compilerError("...");`) rather than a halt() which would also have the benefit of giving the user line number information about where they went off-track.

Reply:
OK, I can make this a compile-time error.